File: GDPIControl/FormMain.vb
Imports System.ComponentModel
Imports System.Net
Imports System.Text.RegularExpressions
Imports GDPIControl.Data

Public Class FormMain
	Private Shared ReadOnly IP_R As New Regex("((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])")
	Private BlockVisible As Boolean
	Private StrToRB As Dictionary(Of String, RadioButton)

	Public Sub New()
		' Этот вызов является обязательным для конструктора.
		InitializeComponent()
		' Добавить код инициализации после вызова InitializeComponent().
		StrToRB = New Dictionary(Of String, RadioButton) From {{"-1", RB_P1}, {"-2", RB_P2}, {"-3", RB_P3}, {"-4", RB_P4}}
		TSMI_Mini.Checked = My.Application.Config.LaunchMinimized
		TSMI_Start.Checked = My.Application.Config.StartOnLaunch

		BlockVisible = My.Application.Config.LaunchMinimized
		If My.Application.Config.StartOnLaunch Then
			StartGDPI()
		Else
			StopGDPI()
		End If
		BS_GDPISettings.DataSource = My.Application.Config.GDPIConfig
		Dim RB = If(StrToRB.ContainsKey(My.Application.Config.Arguments), StrToRB(My.Application.Config.Arguments), RB_Custom)
		RB.Checked = True
	End Sub

	Private Sub B_Copy_Click(sender As Object, e As EventArgs) Handles B_Copy.Click
		If RB_P1.Checked Then
			My.Application.Config.GDPIConfig = GDPISettings.C1
		ElseIf RB_P2.Checked Then
			My.Application.Config.GDPIConfig = GDPISettings.C2
		ElseIf RB_P3.Checked Then
			My.Application.Config.GDPIConfig = GDPISettings.C3
		ElseIf RB_P4.Checked Then
			My.Application.Config.GDPIConfig = GDPISettings.C4
		Else
			Exit Sub
		End If
		BS_GDPISettings.DataSource = My.Application.Config.GDPIConfig
		RB_Custom.Checked = True
	End Sub

	Private Sub StopGDPI()
		B_Start.Text = "Start GDPI"
		TSMI_Enable.Enabled = True
		TSMI_Disable.Enabled = False
		TrayControl.Icon = My.Resources.icon_red
		Icon = My.Resources.icon_red
		GDPIProcess.Stop()
	End Sub

	Private Sub StartGDPI()
		B_Start.Text = "Stop GDPI"
		TSMI_Enable.Enabled = False
		TSMI_Disable.Enabled = True
		TrayControl.Icon = My.Resources.icon_green
		Icon = My.Resources.icon_green
		GDPIProcess.Start()
	End Sub

	Private Sub ExitControl()
		GDPIProcess.Stop()
		TrayControl.Icon = Nothing
		TrayControl.Visible = False
		TrayControl.Dispose()
		Application.Exit()
	End Sub

	Private Sub FormMain_Closing(sender As Object, e As CancelEventArgs) Handles Me.Closing
		e.Cancel = True
		Visible = False
	End Sub

	Private Sub RB_CheckedChanged(sender As Object, e As EventArgs) Handles RB_Custom.CheckedChanged, RB_P1.CheckedChanged, RB_P2.CheckedChanged, RB_P3.CheckedChanged, RB_P4.CheckedChanged
		SetArguments()
		FLP_Custom.Enabled = RB_Custom.Checked
	End Sub

	Private Sub SetArguments()
		Dim Args = ""
		If RB_Custom.Checked Then
			Args = My.Application.Config.GDPIConfig.ToString
		ElseIf RB_P1.Checked Then
			Args = "-1"
		ElseIf RB_P2.Checked Then
			Args = "-2"
		ElseIf RB_P3.Checked Then
			Args = "-3"
		ElseIf RB_P4.Checked Then
			Args = "-4"
		End If
		My.Application.Config.Arguments = Args
	End Sub

	Private Sub ShowForm()
		BlockVisible = False
		Visible = True
	End Sub

#Region "Validating"

	Private Sub DNS_IP_Validating(sender As Object, e As CancelEventArgs) Handles DNS_IP.Validating
		Try
			Dim M = IP_R.Match(DNS_IP.Text)
			e.Cancel = Not M.Success
		Catch ex As Exception
			e.Cancel = True
		End Try
	End Sub

	Private Sub DNS_Port_Validating(sender As Object, e As CancelEventArgs) Handles DNS_Port.Validating
		Try
			Dim p = CInt(DNS_Port.Text)
			e.Cancel = p <= 0 Or p > 65535
		Catch ex As Exception
			e.Cancel = True
		End Try
	End Sub

	Private Sub DNS6_IP_Validating(sender As Object, e As CancelEventArgs) Handles DNS6_IP.Validating
		Try
			Dim IP As IPAddress = Nothing
			Dim r = IPAddress.TryParse(DNS6_IP.Text, IP)
			Dim T = IP.AddressFamily
		Catch ex As Exception
			e.Cancel = True
		End Try
	End Sub

	Private Sub DNS6_Port_Validating(sender As Object, e As CancelEventArgs) Handles DNS6_Port.Validating
		Try
			Dim p = CInt(DNS6_Port.Text)
			e.Cancel = p <= 0 Or p > 65535
		Catch ex As Exception
			e.Cancel = True
		End Try
	End Sub

#End Region

	Private Sub TrayControl_MouseClick(sender As Object, e As MouseEventArgs) Handles TrayControl.MouseClick
		If e.Button = MouseButtons.Left Then ShowForm()
	End Sub

	Private Sub TSMI_Close_Click(sender As Object, e As EventArgs) Handles TSMI_Close.Click
		ExitControl()
	End Sub

	Private Sub TSMI_Settings_Click(sender As Object, e As EventArgs) Handles TSMI_Settings.Click
		ShowForm()
	End Sub

	Private Sub TSMI_Mini_CheckedChanged(sender As Object, e As EventArgs) Handles TSMI_Mini.CheckedChanged
		My.Application.Config.LaunchMinimized = TSMI_Mini.Checked
	End Sub

	Private Sub TSMI_Start_CheckedChanged(sender As Object, e As EventArgs) Handles TSMI_Start.CheckedChanged
		My.Application.Config.StartOnLaunch = TSMI_Start.Checked
	End Sub

	Private Sub TSMI_Enable_Click(sender As Object, e As EventArgs) Handles TSMI_Enable.Click
		StartGDPI()
	End Sub

	Private Sub TSMI_Disable_Click(sender As Object, e As EventArgs) Handles TSMI_Disable.Click
		StopGDPI()
	End Sub

	Private Sub B_Stop_Click(sender As Object, e As EventArgs) Handles B_Stop.Click
		ExitControl()
	End Sub

	Private Sub B_Start_Click(sender As Object, e As EventArgs) Handles B_Start.Click
		If GDPIProcess.IsRunning Then
			StopGDPI()
		Else
			StartGDPI()
		End If
	End Sub

	Protected Overrides Sub SetVisibleCore(value As Boolean)
		If BlockVisible Then Exit Sub
		MyBase.SetVisibleCore(value)
	End Sub

	Protected Overrides Sub OnLoad(e As EventArgs)
		MyBase.OnLoad(e)
	End Sub

	Private Sub TSMI_Blacklist_Click(sender As Object, e As EventArgs) Handles TSMI_Blacklist.Click
		Dim F = New FormBlacklist
		F.ShowDialog(Me)
	End Sub

	Private Sub TSMI_Userlist_Click(sender As Object, e As EventArgs) Handles TSMI_Userlist.Click
		Process.Start("Userlist.txt")
	End Sub

End Class

File: GDPIControl/FormMain.cs
using GDPIControl.Extensions;
using GDPIControl.Forms;
using GDPIControl.Model;
using GDPIControl.Properties;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace GDPIControl
{
    public partial class FormMain : Form
    {
        private readonly HashSet<Modeset> CustomModesets;
        private readonly List<(RadioButton Button, Modeset Modeset)> RBModesets;
        private readonly ControlSettings Settings = Config.Current;

        public FormMain()
        {
            InitializeComponent();
            this.BindSettings();
            CustomModesets = new() {
                Modeset.Custom1,
                Modeset.Custom2,
                Modeset.Custom3
            };
            RBModesets = new List<(RadioButton Button, Modeset Modeset)>
            {
                (RB_M1,Modeset.M1),
                (RB_M2,Modeset.M2),
                (RB_M3,Modeset.M3),
                (RB_M4,Modeset.M4),
                (RB_M5,Modeset.M5),
                (RB_M6,Modeset.M6),
                (RB_M7,Modeset.M7),
                (RB_M8,Modeset.M8),
                (RB_M9,Modeset.M9),
                (RB_Custom_1,Modeset.Custom1),
                (RB_Custom_2,Modeset.Custom2),
                (RB_Custom_3,Modeset.Custom3)
            };
            RBModesets.First(X => X.Modeset == Settings.Modeset).Button.Checked = true;
            BS_ControlSettings.DataSource = Settings;

            RefreshUI();
            RefreshArguments();
            if (Settings.AutostartGDPI) { StartGDPI(); } else { StopGDPI(); }
        }

        private void CloseGDPIControl()
        {
            GDPIProcess.Stop();
            TrayControl.Icon = null;
            TrayControl.Visible = false;
            TrayControl.Dispose();
            RefreshArguments();
            if (Settings.StopDriver)
            {
                GDPIProcess.StopDriver();
            }
            Application.Exit();
        }

        private void RefreshArguments()
        {
            Settings.Modeset = RBModesets.First(X => X.Button.Checked).Modeset;
            var Arguments = Settings.Modeset.ToArgument();
            if (Settings.UseBlacklist) { Arguments += $@" --blacklist ""{Constants.BlacklistPath}"""; }
            if (Settings.UseUserlist) { Arguments += $@" --blacklist ""{Constants.UserlistPath}"""; }
            Settings.Arguments = Arguments;
        }

        private void RefreshUI()
        {
            B_Edit.Enabled = CustomModesets.Contains(Settings.Modeset);
            B_Restart.Enabled = GDPIProcess.IsRunning;
        }

        private void ShowGDPIControl()
        {
            Show();
            Activate();
        }

        private void StartGDPI()
        {
            B_Start.Text = "Stop GDPI";
            B_Start.Image = Resources.GDPI_red;
            MI_Start.Enabled = false;
            MI_Stop.Enabled = true;
            TrayControl.Icon = Resources.icon_green;
            Icon = Resources.icon_green;
            RefreshArguments();
            GDPIProcess.Start();
            RefreshUI();
        }

        private void StopGDPI()
        {
            B_Start.Text = "Start GDPI";
            B_Start.Image = Resources.GDPI_green;
            MI_Start.Enabled = true;
            MI_Stop.Enabled = false;
            TrayControl.Icon = Resources.icon_red;
            Icon = Resources.icon_red;
            RefreshArguments();
            GDPIProcess.Stop();
            RefreshUI();
        }

        private void UIState(bool state)
        {
            B_Restart.Enabled = state;
            B_Start.Enabled = state;
            B_Close.Enabled = state;
        }

        #region UIEvents

        private void B_Close_Click(object sender, EventArgs e) => CloseGDPIControl();

        private void B_Edit_Click(object sender, EventArgs e)
        {
            using var form = new FormGDPISettings(Settings);
            if (form.ShowDialog(this) == DialogResult.OK)
            {
                RefreshArguments();
            }
        }

        private async void B_Restart_Click(object sender, EventArgs e)
        {
            UIState(false);
            StopGDPI();
            await Task.Delay(500);
            StartGDPI();
            UIState(true);
        }

        private void B_Start_Click(object sender, EventArgs e)
        {
            UIState(false);
            if (GDPIProcess.IsRunning) { StopGDPI(); } else { StartGDPI(); }
            UIState(true);
        }

        private void FormMain_FormClosing(object sender, FormClosingEventArgs e)
        {
            if (e.CloseReason == CloseReason.ApplicationExitCall) { return; }
            e.Cancel = true;
            Visible = false;
        }

        private void MI_About_Click(object sender, EventArgs e)
        {
            var F = new FormAbout();
            F.ShowDialog(this);
        }

        private void MI_Settings_Click(object sender, EventArgs e)
        {
            using var F = new FormSettings();
            F.ShowDialog(this);
        }

        private void RB_CheckedChanged(object sender, EventArgs e)
        {
            var RB = (RadioButton)sender;
            if (!RB.Checked) { return; }
            foreach (var entry in RBModesets)
            {
                if (entry.Button == RB) { continue; }
                entry.Button.Checked = false;
            }
            RefreshArguments();
            RefreshUI();
        }

        #region Lists

        private void MI_Blacklist_Click(object sender, EventArgs e)
        {
            var F = new FormBlacklist();
            F.ShowDialog(this);
        }

        private void MI_Userlist_Click(object sender, EventArgs e)
        {
            if (!File.Exists(Constants.UserlistPath))
            {
                File.WriteAllText(Constants.UserlistPath, "");
            }
            Process.Start(new ProcessStartInfo(Constants.UserlistPath) { UseShellExecute = true });
        }

        #endregion Lists

        #region Tray

        private void MI_Close_Click(object sender, EventArgs e) => CloseGDPIControl();

        private void MI_Show_Click(object sender, EventArgs e) => ShowGDPIControl();

        private void MI_Start_Click(object sender, EventArgs e) => StartGDPI();

        private void MI_Stop_Click(object sender, EventArgs e) => StopGDPI();

        private void TrayControl_MouseClick(object sender, MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Left) { ShowGDPIControl(); }
        }

        #endregion Tray

        #endregion UIEvents
    }
}